Issue #1076: Add ability to add an existing node to a layout as a block

docwilmot opened the pull request:
Fixes https://github.com/backdrop/backdrop-issues/issues/1076

docwilmot commented:
reopening for sandbox

docwilmot commented:
@gifad thanks for the review; youre right again about the `$node->title` being set to NULL; if a block _with_ the title followed a block _without_, both show no title as you warned if they are both of the same node. Unlikely use case, but you never know. Have a look at this new effort please.
- cloned the node as recommended
- removed the 'clone_title' property as it isnt needed (its still in this PR, because I forgot to delete it, but its not used)
- also changed from NULL to empty string as well; probably safer as nodes should have titles
- added a view mode called block (excellent idea, yes) with no author details (not sure about that part, maybe it should be kept).

Example in the sandbox shows node with title following node without. Would appreciate your review again.

ghost commented:
> Would appreciate your review again.

Thanks, the feature is really nice.
To be perfect, two points may still need work, IMHO : 
1. `+  if ($node_type->settings['node_submitted'] && $variables['view_mode'] != 'block') {`
   I disagree : _submitted_ is frequently used in a "featured content" like block, at top right of a home page.
   _submitted_ is often requested here, and one would not use _teaser_ mode, as _teaser_ also shows an image, or whatever reason.
   So, either keep with the default value, or provide yet another setting for the block…
2. Some themes (bartik first), show the node title in a h2 tag in a link.
   If the title is empty, we still have a 
   `<a href="/node/1"></a>`
   line, which, combined with

```
.l-sidebar h2 {
    border-bottom: 1px solid #d6d6d6;
```

leads with an unexpected horizontal bar in the block.
This cannot be qualified as a bartik bug, since node officialy cannot be empty.
It should be addressed in your PR, though.
I'm not confident in `$variable['page']`, so I could not find better than, in a hook_preprocess_node() : 

```
    if (isset($variables['title']) && empty($variables['title'])) {
    // hide the empty <h2>
      $variables['title_prefix'] = '<div style="display: none;">';
      $variables['title_suffix'] = '</div>';
    }
```

docwilmot commented:
Thanks @gifad 
Re 1, agreed, I wasnt sure about that either. will delete.
Re 2, also agree we should address the empty title text. I partly addressed this in the default `node.tpl.php` (see PR) by adding a check for `empty($title)`:

```
<?php if (!$page && !empty($title)): ?>
  <h2><a href="<?php print $node_url; ?>" rel="bookmark"><?php print $title; ?></a></h2>
<?php endif; ?>
```

and would propose to modify Bartik's `node.tpl.php` similarly, just didnt do it in this PR yet. I think if we are adding the new concept of title-less nodes, we should add a more robust check for missing titles in core and core themes. I dont think just hiding it with CSS is enough.

What would be best though is if there was some way to add an extra option in the block form for hiding the `submitted_by` stuff, but I can't figure that out how to access this on block display. Toggling `submitted_by` is done at the node type form, and if thats set the whole text block is built in `node.theme.inc`. But there is no context in `node.theme.inc` to know if the node is being shown in a block, so no access to block settings. I could in `ContentBlock::getContent()` add a key to the `content` array like `$content['show_submitted'] = TRUE` which would be available in `node.theme.inc` but not sure of the ramifications of that.

Any thoughts?

ghost commented:
> I can't figure that out how to access this on block display

we are in a block, but we are also in a node; so what do you think of : 

```
hook_preprocess_node(&$variables) {
…
      $variables['display_submitted'] = $settings['display_submitted'];
```

`<aside>` This is not closeley related, but I just noted that if you make the main content block style "dynamic", it gets tagged with `class="block"`, while default style, it is not.
Then bartik apply 

```
.block ol, .block ul {
  margin: 0;
```

which takes precedence on 

```
  .vertical-tabs-list {
    margin: -1px 0 -1px -15em;
```

just an example ...

Hence my inelegant and inefficient, but safe, local preprocess workaround...
(given my low level as a drupal/backdrop core developper, I prefer "think global, act local" !)

docwilmot commented:
> $variables['display_submitted'] = $settings['display_submitted'];

Problem is `$settings` would not be available in `hook_preprocess_node` (unless as I mentioned we pass this through `ContentBlock::getContent()`. Is that what you mean?

ghost commented:
Talking about low level, my project [nice_blocks](https://github.com/backdrop-contrib/nice_blocks) is a joke, but it works...
![capture d ecran 2016-03-31 a 00 22 08](https://cloud.githubusercontent.com/assets/15943462/14159608/f456ad98-f6d6-11e5-82b8-a6cc5283bae3.jpg)

docwilmot commented:
Updated. Can now hide "submitted by" info, and also by default hides the title div if no title, both in the default `node.tpl.php` and in Bartik.

Shoout to @quicksketch @jenlampton @mikemccaffrey I added a "block" view mode here to core as part of this PR, but I'm now having doubts, as it is essentially the sane as the "full' view mode, with no intrinsic difference in display. Please advise if I should keep this.

jenlampton commented:
Great work @docwilmot!  I would say let's not add another view mode. If people choose to add one on their own sites that's fine, but in core let's keep it simple :)

ghost commented:
@jenlampton 

> If people choose to add one

Sorry for the (probably) dumb question, but where is the UI people can use ?
Can't remember anything other than a custom module…

quicksketch commented:
> Sorry for the (probably) dumb question, but where is the UI people can use? Can't remember anything other than a custom module…

You're right, currently you cannot make new display modes without a contrib or custom module. So for now people will either need to use teaser or full display modes. Either adding new display modes or making it possible to add custom ones should be a separate issue.

quicksketch commented on core/modules/node/node.content_block.inc:
```diff
@@ -0,0 +1,163 @@
+ *  
+ *  This class allows us to display a particular node in a block.
+ */
+class ContentBlock extends Block {
```
Although we use "Content" on the front-end, sticking to our convention of "Node" is good for namespacing this class. Could we rename to `NodeBlock` or `NodeContentBlock`? If choosing `NodeBlock`, the file name may also be shorted to `node.block.inc`, but if we go with `NodeContentBlock` it can stay as-is.

jenlampton commented:
> where is the UI people can use

There is a UI in the entity view mode module (and there is an issue to add that into core for Backdrop see https://github.com/backdrop/backdrop-issues/issues/315). But I usually use a custom module myself :)

I have a new PR coming for this where I am turning this into an entity block that can be extended by any entity.

docwilmot commented:
Replaced by https://github.com/backdrop/backdrop/pull/1529
